### DISCUSSION TITLE
Get PropMap from RefArg

### DISCUSSION THREAD

**Discussion by hpux735** [diwic/dbus-rs#326] [Q&A] [answered]
I've been working on decoding a dbus like the following:

`Got message: Message { Type: Signal, Path: "/com/racepointenergy/CLIAgent1", Interface: "com.racepointenergy.RPEBus1", Member: "network", Sender: ":1.3", Serial: 872, Args: [{"contents": Variant({"status": Variant("connected"), "ipAddress": Variant("127.0.0.1"), "port": Variant(2000)}), "regarding": Variant("network"), "timestamp": Variant("2021-03-24T20:21:53.280Z")}] }`

I've gotten it far enough that I have a RefArg for the "contents" argument.  But the nested HashMap is giving me trouble.  I've gotten to the point where if I print (using `{:?}`) I get `{"port": Variant(2000), "ipAddress": Variant("127.0.0.1"), "status": Variant("connected")}`, but the only thing I've been able to get out of that is an iterator, where the keys and values alternate in the iterator. For example:

```
                let value = &args["contents"].0;
                println!("{:?}", value);
                if let Some(a) = value.as_iter() {
                    for i in a {
                        println!("{:?}", i);
                    }
                }
```
yields:
```
{"port": Variant(2000), "ipAddress": Variant("127.0.0.1"), "status": Variant("connected")}
"port"
Variant(2000)
"ipAddress"
Variant("127.0.0.1")
"status"
Variant("connected")
```

How can I get a HashMap instead?  The documentation says:

> Dicts and variants are sometimes combined, e g, you might need to read a D-Bus dictionary of String to Variants. You can read these as PropMap (which is a type alias for HashMap<String, Variant<Box<dyn RefArg>>>) and use prop_cast to retrieve a value.

But, that's not really that clear to someone not steeped in this library.  How can I read these as a PropMap?  How do I use prop_cast?  These things aren't in an example that I can find.

Thanks!

**Comment by diwic** [comment, marked as answer]
Hmm, there might be more than one way depending on the actual signature of the signal's argument. Does this work?

    let value = &args["contents"].0;
    let map: &arg::PropMap = arg::cast(value)?;

If not, could you show the introspection XML (or some other information) about the signal?

**Reply by hpux735** [reply]
YES!!  That does work!  Thank you!
